storage/remote/queue_manager.go

```diff
@@ -424,22 +476,30 @@ func (s *shards) stop(deadline time.Duration) {
 
 	// Force an unclean shutdown.
 	s.cancel()
-	<-s.done
+	// Why does this send block now...?
+	// <-s.done
 	return
 }
 
-func (s *shards) enqueue(sample *model.Sample) bool {
+func (s *shards) enqueue(sample tsdb.RefSample) bool {
 	s.qm.samplesIn.incr(1)
 
-	fp := sample.Metric.FastFingerprint()
-	shard := uint64(fp) % uint64(len(s.queues))
+	// fp := sample.Metric.FastFingerprint()
+	shard := uint64(sample.Ref) % uint64(len(s.queues))
 
-	select {
-	case s.queues[shard] <- sample:
-		return true
-	default:
+	if int(shard) >= len(s.queues) {
+		level.Error(s.qm.logger).Log("err", "we would be chosing a non-existant shard", "shard", shard, "numQueues", len(s.queues))
+	}
+
+	if len(s.queues[shard]) == cap(s.queues[shard]) {
+		if s.qm.logLimiter.Allow() {
+			// This still seems to log frequently.
+			level.Info(s.qm.logger).Log("msg", "Shard queue is full. Multiple subsequent messages of this kind may be suppressed.", "shard", shard)
+		}
 		return false
 	}
```

Comment:
No logging here pls: far to noisy.  Shards being full is common case and is out back pressure mechanism.